File: ios/NitroOTAModule/Core/OtaManager.swift
//
//  OtaManager.swift
//  NitroOta
//
//  Created by Ritesh Shukla on 19/10/25.
//


import Foundation

class OtaManager {
    private let downloadManager = DownloadManager()
    private let preferences = PreferencesUtils()
    private let androidBundleName = "main.jsbundle"

    // MARK: - Private helper methods

    private func cleanupOldOtaDirectories() {
        do {
            let filesDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            let otaDirectories = try FileManager.default.contentsOfDirectory(at: filesDir, includingPropertiesForKeys: [.isDirectoryKey], options: [])

            let filteredDirs = otaDirectories.filter { url in
                do {
                    let resourceValues = try url.resourceValues(forKeys: [.isDirectoryKey])
                    return resourceValues.isDirectory ?? false && url.lastPathComponent.hasPrefix("ota_unzipped_")
                } catch {
                    return false
                }
            }

            if filteredDirs.count <= 2 {
                print("OtaManager: Found \(filteredDirs.count) OTA directories, no cleanup needed")
                return
            }

            // Sort by timestamp (newest first) - extract timestamp from folder name
            let sortedDirs = filteredDirs.sorted { (url1, url2) -> Bool in
                let timestampStr1 = url1.lastPathComponent.replacingOccurrences(of: "ota_unzipped_", with: "")
                let timestampStr2 = url2.lastPathComponent.replacingOccurrences(of: "ota_unzipped_", with: "")

                let timestamp1 = Int(timestampStr1) ?? 0
                let timestamp2 = Int(timestampStr2) ?? 0

                return timestamp1 > timestamp2
            }

            // Keep only the first 2 (most recent), delete the rest
            let dirsToDelete = Array(sortedDirs.dropFirst(2))
            print("OtaManager: Found \(filteredDirs.count) OTA directories, keeping 2 most recent, deleting \(dirsToDelete.count) old ones")

            for dir in dirsToDelete {
                do {
                    try FileManager.default.removeItem(at: dir)
                    print("OtaManager: Successfully deleted old OTA directory: \(dir.path)")
                } catch {
                    print("OtaManager: Error deleting old OTA directory \(dir.path): \(error.localizedDescription)")
                }
            }
        } catch {
            print("OtaManager: Error during OTA directory cleanup: \(error.localizedDescription)")
        }
    }

    private func findAndRenameContentFolder(unzipDir: URL) -> URL? {
        do {
            let contents = try FileManager.default.contentsOfDirectory(at: unzipDir, includingPropertiesForKeys: [.isDirectoryKey], options: [])

            for item in contents {
                do {
                    let resourceValues = try item.resourceValues(forKeys: [.isDirectoryKey])
                    if resourceValues.isDirectory ?? false {
                        // Found a directory, rename it to "bundles"
                        let bundlesDir = unzipDir.appendingPathComponent("bundles")
                        try FileManager.default.moveItem(at: item, to: bundlesDir)
                        print("OtaManager: Renamed content folder from '\(item.lastPathComponent)' to 'bundles'")
                        return bundlesDir
                    }
                } catch {
                    print("OtaManager: Error checking item \(item.path): \(error.localizedDescription)")
                }
            }
        } catch {
            print("OtaManager: Error reading unzip directory contents: \(error.localizedDescription)")
        }

        return nil
    }

    private func readVersionFromLocalFile(_ contentFolder: URL) {
        let otaVersionFile = contentFolder.appendingPathComponent("ota.version")
        do {
            let otaVersion = try String(contentsOf: otaVersionFile, encoding: .utf8).trimmingCharacters(in: .whitespacesAndNewlines)
            print("OtaManager: Found ota.version file with content: \(otaVersion)")
            preferences.setOtaVersion(otaVersion)
            print("OtaManager: Stored OTA version in UserDefaults: \(otaVersion)")
        } catch {
            print("OtaManager: ota.version file not found in content folder: \(contentFolder.path)")
        }
    }

    private func downloadVersionFromUrl(_ versionUrl: String) throws -> String {
        print("OtaManager: Downloading version from URL: \(versionUrl)")
        let versionContent = try downloadManager.downloadText(from: versionUrl).trimmingCharacters(in: .whitespacesAndNewlines)
        print("OtaManager: Downloaded version: \(versionContent)")
        return versionContent
    }

    // MARK: - Public methods

    func downloadAndUnzipFromUrl(_ downloadUrl: String, versionCheckUrl: String?) throws -> String {
        // Store the URLs for future update checks
        preferences.setUpdateDownloadUrl(downloadUrl)
        if let versionCheckUrl = versionCheckUrl {
            preferences.setUpdateVersionCheckUrl(versionCheckUrl)
        }
        print("OtaManager: Stored download URL: \(downloadUrl) and version check URL: \(versionCheckUrl ?? "nil")")

        // Create a temporary file for the zip
        let documentsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let zipFile = documentsDir.appendingPathComponent("ota_update_\(UUID().uuidString).zip")
        print("OtaManager: Created temp zip file: \(zipFile.path)")

        do {
            // Download the zip file
            guard let downloadURL = URL(string: downloadUrl) else {
                throw NSError(domain: "OtaManager", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid download URL"])
            }
            print("OtaManager: Downloading zip file...")
            try downloadManager.downloadFile(from: downloadURL, to: zipFile)
            print("OtaManager: Download completed, zip file size: \(try? FileManager.default.attributesOfItem(atPath: zipFile.path)[.size] as? NSNumber ?? 0) bytes")

            // Create directory for unzipped files
            let unzipDir = documentsDir.appendingPathComponent("ota_unzipped_\(Int(Date().timeIntervalSince1970 * 1000))")
            try FileManager.default.createDirectory(at: unzipDir, withIntermediateDirectories: true, attributes: nil)
            print("OtaManager: Created unzip directory: \(unzipDir.path)")

            // Unzip the file
            print("OtaManager: Starting unzip process...")
            try ZipUtils.unzip(zipFile: zipFile, to: unzipDir)
            let contents = try? FileManager.default.contentsOfDirectory(atPath: unzipDir.path)
            print("OtaManager: Unzip completed, directory contents: \(contents?.joined(separator: ", ") ?? "No items")")

            // Find the actual content folder and rename it to "bundles"
            let contentFolder = findAndRenameContentFolder(unzipDir: unzipDir)
            if let contentFolder = contentFolder {
                print("OtaManager: Using content folder: \(contentFolder.path)")

                // Read version from provided URL or fallback to ota.version file
                if let versionCheckUrl = versionCheckUrl {
                    do {
                        let otaVersion = try downloadVersionFromUrl(versionCheckUrl)
                        print("OtaManager: Downloaded version from URL: \(otaVersion)")
                        preferences.setOtaVersion(otaVersion)
                        print("OtaManager: Stored OTA version in UserDefaults: \(otaVersion)")
                    } catch {
                        print("OtaManager: Failed to download version from URL: \(versionCheckUrl), error: \(error.localizedDescription)")
                        // Fallback to local file
                        readVersionFromLocalFile(contentFolder)
                    }
                } else {
                    // No version check URL provided, try to read from local file
                    readVersionFromLocalFile(contentFolder)
                }

                // Store the content folder path in user defaults
                preferences.setOtaUnzippedPath(contentFolder.appendingPathComponent(androidBundleName).path)
                print("OtaManager: Stored content folder path in UserDefaults: \(contentFolder.path)")

                // Store the Android bundle name in user defaults
                preferences.setOtaBundleName(androidBundleName)
                print("OtaManager: Stored Android bundle name in UserDefaults: \(androidBundleName)")

                // Clean up old OTA directories, keeping only the 2 most recent
                cleanupOldOtaDirectories()

                return contentFolder.path
            } else {
                print("OtaManager: No content folder found in unzip directory")
                // Fallback to unzip directory
                preferences.setOtaUnzippedPath(unzipDir.path)
                print("OtaManager: Stored fallback unzip path in UserDefaults: \(unzipDir.path)")

                // Try to read version from fallback directory if versionCheckUrl provided
                if let versionCheckUrl = versionCheckUrl {
                    do {
                        let otaVersion = try downloadVersionFromUrl(versionCheckUrl)
                        preferences.setOtaVersion(otaVersion)
                        print("OtaManager: Stored OTA version from URL in UserDefaults: \(otaVersion)")
                    } catch {
                        print("OtaManager: Failed to download version from URL for fallback: \(versionCheckUrl), error: \(error.localizedDescription)")
                    }
                }

                // Store the Android bundle name in user defaults
                preferences.setOtaBundleName(androidBundleName)
                print("OtaManager: Stored Android bundle name in UserDefaults: \(androidBundleName)")

                // Clean up old OTA directories, keeping only the 2 most recent
                cleanupOldOtaDirectories()

                return unzipDir.path
            }
        } catch {
            print("OtaManager: Error in downloadAndUnzipFromUrl for URL: \(downloadUrl), error: \(error.localizedDescription)")
            throw error
        }

        // Always delete the zip file
        do {
            try FileManager.default.removeItem(at: zipFile)
            print("OtaManager: Cleaned up temp zip file: \(zipFile.path)")
        } catch {
            print("OtaManager: Failed to clean up temp zip file: \(zipFile.path), error: \(error.localizedDescription)")
        }
    }

    func getStoredUnzippedPath() -> String? {
        return preferences.getOtaUnzippedPath()
    }

    func getStoredBundleName() -> String? {
        return preferences.getOtaBundleName()
    }

    func getStoredOtaVersion() -> String? {
        return preferences.getOtaVersion()
    }

    func checkForUpdates(_ versionCheckUrl: String? = nil) throws -> Bool {
        let checkUrl = try versionCheckUrl ?? preferences.getUpdateVersionCheckUrl() ?? {
            throw NSError(domain: "OtaManager", code: -1, userInfo: [NSLocalizedDescriptionKey: "No version check URL provided and none stored. Call downloadAndUnzipFromUrl with versionCheckUrl first or provide URL."])
        }()

        let storedVersion = preferences.getOtaVersion()

        print("OtaManager: Checking for updates using version check URL: \(checkUrl)")
        print("OtaManager: Current stored version: \(storedVersion ?? "nil")")

        // Download current version from the URL
        let currentVersion = try downloadVersionFromUrl(checkUrl)
        print("OtaManager: Latest version from URL: \(currentVersion)")

        // Compare versions
        let hasUpdate = storedVersion != currentVersion

        print("OtaManager: Update available: \(hasUpdate)")

        return hasUpdate
    }

    func clearStoredData() {
        preferences.clearOtaData()
        print("OtaManager: Cleared stored OTA data from UserDefaults")
    }
}

File: react-native-nitro-ota/android/src/main/java/com/margelo/nitro/nitroota/core/OtaManager.kt
package com.margelo.nitro.nitroota.core

import android.content.Context
import com.margelo.nitro.nitroota.network.DownloadManager
import com.margelo.nitro.nitroota.utils.PreferencesUtils
import com.margelo.nitro.nitroota.utils.UrlUtils
import com.margelo.nitro.nitroota.utils.ZipUtils
import java.io.File
import android.util.Log

private const val ANDROID_BUNDLE_NAME = "index.android.bundle"


class OtaManager(
    private val context: Context,
    private val downloadManager: DownloadManager = DownloadManager(),
    private val preferences: PreferencesUtils = PreferencesUtils.create(context)
) {

    /**
     * Cleans up old OTA unzip directories, keeping only the 2 most recent ones.
     * This ensures we don't accumulate too many old update folders.
     */
    private fun cleanupOldOtaDirectories() {
        try {
            val filesDir = context.filesDir
            val otaDirectories = filesDir.listFiles { file ->
                file.isDirectory && file.name.startsWith("ota_unzipped_")
            }?.toList() ?: emptyList()

            if (otaDirectories.size <= 2) {
                Log.d("OtaManager", "Found ${otaDirectories.size} OTA directories, no cleanup needed")
                return
            }

            // Sort by timestamp (newest first) - extract timestamp from folder name
            val sortedDirs = otaDirectories.sortedByDescending { dir ->
                try {
                    // Extract timestamp from "ota_unzipped_<timestamp>" format
                    val timestampStr = dir.name.substringAfter("ota_unzipped_")
                    timestampStr.toLong()
                } catch (e: Exception) {
                    // If parsing fails, treat as very old (timestamp 0)
                    Log.w("OtaManager", "Failed to parse timestamp from directory name: ${dir.name}", e)
                    0L
                }
            }

            // Keep only the first 2 (most recent), delete the rest
            val dirsToDelete = sortedDirs.drop(2)
            Log.d("OtaManager", "Found ${otaDirectories.size} OTA directories, keeping 2 most recent, deleting ${dirsToDelete.size} old ones")

            dirsToDelete.forEach { dir ->
                try {
                    val deleted = dir.deleteRecursively()
                    if (deleted) {
                        Log.d("OtaManager", "Successfully deleted old OTA directory: ${dir.absolutePath}")
                    } else {
                        Log.w("OtaManager", "Failed to delete old OTA directory: ${dir.absolutePath}")
                    }
                } catch (e: Exception) {
                    Log.e("OtaManager", "Error deleting old OTA directory: ${dir.absolutePath}", e)
                }
            }
        } catch (e: Exception) {
            Log.e("OtaManager", "Error during OTA directory cleanup", e)
        }
    }

    /**
     * Finds the content folder inside the unzip directory and renames it to "bundles".
     * GitHub zips typically create a folder with the format "repo-name-branch".
     *
     * @param unzipDir The directory where files were unzipped
     * @return The renamed content folder, or null if not found
     */
    private fun findAndRenameContentFolder(unzipDir: File): File? {
        val files = unzipDir.listFiles()
        if (files != null) {
            for (file in files) {
                if (file.isDirectory) {
                    // Found a directory, rename it to "bundles"
                    val bundlesDir = File(unzipDir, "bundles")
                    val renamed = file.renameTo(bundlesDir)
                    if (renamed) {
                        Log.d("OtaManager", "Renamed content folder from '${file.name}' to 'bundles'")
                        return bundlesDir
                    } else {
                        Log.w("OtaManager", "Failed to rename folder '${file.name}' to 'bundles'")
                        return file // Return the original folder if rename failed
                    }
                }
            }
        }
        return null
    }

    /**
     * Downloads a zip file from any URL, unzips it, and stores the path.
     *
     * @param downloadUrl The direct URL to download the zip file from
     * @param versionCheckUrl The URL to check for version information (optional)
     * @return The absolute path to the unzipped directory
     * @throws Exception If download or unzip fails
     */
    fun downloadAndUnzipFromUrl(downloadUrl: String, versionCheckUrl: String? = null): String {
        // Store the URLs for future update checks
        preferences.setUpdateDownloadUrl(downloadUrl)
        if (versionCheckUrl != null) {
            preferences.setUpdateVersionCheckUrl(versionCheckUrl)
        }
        Log.d("OtaManager", "Stored download URL: $downloadUrl and version check URL: $versionCheckUrl")

        // Create a temporary file for the zip
        val zipFile = File.createTempFile("ota_update", ".zip", context.filesDir)
        Log.d("OtaManager", "Created temp zip file: ${zipFile.absolutePath}")

        try {
            // Download the zip file
            Log.d("OtaManager", "Downloading zip file...")
            downloadManager.downloadFile(downloadUrl, zipFile)
            Log.d("OtaManager", "Download completed, zip file size: ${zipFile.length()} bytes")

            // Create directory for unzipped files
            val unzipDir = File(context.filesDir, "ota_unzipped_${System.currentTimeMillis()}")
            unzipDir.mkdirs()
            Log.d("OtaManager", "Created unzip directory: ${unzipDir.absolutePath}")

            // Unzip the file
            Log.d("OtaManager", "Starting unzip process...")
            ZipUtils.unzip(zipFile, unzipDir)
            Log.d("OtaManager", "Unzip completed, directory contents: ${unzipDir.list()?.joinToString(", ") ?: "No items"}")

            // Find the actual content folder and rename it to "bundles"
            val contentFolder = findAndRenameContentFolder(unzipDir)
            if (contentFolder != null) {
                Log.d("OtaManager", "Using content folder: ${contentFolder.absolutePath}")

                // Read version from provided URL or fallback to ota.version file
                if (versionCheckUrl != null) {
                    try {
                        val otaVersion = downloadVersionFromUrl(versionCheckUrl)
                        Log.d("OtaManager", "Downloaded version from URL: $otaVersion")
                        preferences.setOtaVersion(otaVersion)
                        Log.d("OtaManager", "Stored OTA version in SharedPreferences: $otaVersion")
                    } catch (e: Exception) {
                        Log.w("OtaManager", "Failed to download version from URL: $versionCheckUrl", e)
                        // Fallback to local file
                        readVersionFromLocalFile(contentFolder)
                    }
                } else {
                    // No version check URL provided, try to read from local file
                    readVersionFromLocalFile(contentFolder)
                }

                // Store the content folder path in shared preferences
                preferences.setOtaUnzippedPath(contentFolder.absolutePath + "/" + ANDROID_BUNDLE_NAME)
                Log.d("OtaManager", "Stored content folder path in SharedPreferences: ${contentFolder.absolutePath}")

                // Store the Android bundle name in shared preferences
                preferences.setOtaBundleName(ANDROID_BUNDLE_NAME)
                Log.d("OtaManager", "Stored Android bundle name in SharedPreferences: $ANDROID_BUNDLE_NAME")

                // Clean up old OTA directories, keeping only the 2 most recent
                cleanupOldOtaDirectories()

                return contentFolder.absolutePath
            } else {
                Log.w("OtaManager", "No content folder found in unzip directory")
                // Fallback to unzip directory
                preferences.setOtaUnzippedPath(unzipDir.absolutePath)
                Log.d("OtaManager", "Stored fallback unzip path in SharedPreferences: ${unzipDir.absolutePath}")

                // Try to read version from fallback directory if versionCheckUrl provided
                if (versionCheckUrl != null) {
                    try {
                        val otaVersion = downloadVersionFromUrl(versionCheckUrl)
                        preferences.setOtaVersion(otaVersion)
                        Log.d("OtaManager", "Stored OTA version from URL in SharedPreferences: $otaVersion")
                    } catch (e: Exception) {
                        Log.w("OtaManager", "Failed to download version from URL for fallback: $versionCheckUrl", e)
                    }
                }

                // Store the Android bundle name in shared preferences
                preferences.setOtaBundleName(ANDROID_BUNDLE_NAME)
                Log.d("OtaManager", "Stored Android bundle name in SharedPreferences: $ANDROID_BUNDLE_NAME")

                // Clean up old OTA directories, keeping only the 2 most recent
                cleanupOldOtaDirectories()

                return unzipDir.absolutePath
            }
        } catch (e: Exception) {
            Log.e("OtaManager", "Error in downloadAndUnzipFromUrl for URL: $downloadUrl", e)
            throw e // Re-throw the exception
        } finally {
            // Always delete the zip file
            val deleted = zipFile.delete()
            Log.d("OtaManager", "Cleaned up temp zip file: ${zipFile.absolutePath}, deleted: $deleted")
        }
    }

    /**
     * Gets the stored unzipped path from shared preferences.
     *
     * @return The stored unzipped directory path, or null if not found
     */
    fun getStoredUnzippedPath(): String? {
        return preferences.getOtaUnzippedPath()
    }

    /**
     * Gets the stored Android bundle name from shared preferences.
     *
     * @return The stored Android bundle name, or null if not found
     */
    fun getStoredBundleName(): String? {
        return preferences.getOtaBundleName()
    }

    /**
     * Gets the stored OTA version from shared preferences.
     *
     * @return The stored OTA version, or null if not found
     */
    fun getStoredOtaVersion(): String? {
        return preferences.getOtaVersion()
    }

    /**
     * Checks for updates by downloading the current version from the version check URL
     * and comparing it with the stored version. Returns true if a new version is available.
     *
     * @param versionCheckUrl The URL to check for version information (optional, uses stored URL if not provided)
     * @return true if a new version is available, false if current version is up to date
     * @throws Exception If download fails
     */
    fun checkForUpdates(versionCheckUrl: String? = null): Boolean {
        val checkUrl = versionCheckUrl ?: preferences.getUpdateVersionCheckUrl()
            ?: throw IllegalStateException("No version check URL provided and none stored. Call downloadAndUnzipFromUrl with versionCheckUrl first or provide URL.")

        val storedVersion = preferences.getOtaVersion()

        Log.d("OtaManager", "Checking for updates using version check URL: $checkUrl")
        Log.d("OtaManager", "Current stored version: $storedVersion")

        // Download current version from the URL
        val currentVersion = downloadVersionFromUrl(checkUrl)
        Log.d("OtaManager", "Latest version from URL: $currentVersion")

        // Compare versions
        val hasUpdate = storedVersion != currentVersion

        Log.d("OtaManager", "Update available: $hasUpdate")

        return hasUpdate
    }

    /**
     * Reads the ota.version file from the local content folder.
     *
     * @param contentFolder The folder to read the version file from
     */
    private fun readVersionFromLocalFile(contentFolder: File) {
        // Read ota.version file from the content folder
        val otaVersionFile = File(contentFolder, "ota.version")
        if (otaVersionFile.exists() && otaVersionFile.isFile) {
            val otaVersion = otaVersionFile.readText().trim()
            Log.d("OtaManager", "Found ota.version file with content: $otaVersion")
            preferences.setOtaVersion(otaVersion)
            Log.d("OtaManager", "Stored OTA version in SharedPreferences: $otaVersion")
        } else {
            Log.w("OtaManager", "ota.version file not found in content folder: ${contentFolder.absolutePath}")
        }
    }

    /**
     * Downloads the version from any URL and returns its content.
     *
     * @param versionUrl The URL to download the version from
     * @return The content of the version file/endpoint
     * @throws Exception If download fails
     */
    private fun downloadVersionFromUrl(versionUrl: String): String {
        Log.d("OtaManager", "Downloading version from URL: $versionUrl")

        // Download the version content as text
        val versionContent = downloadManager.downloadText(versionUrl).trim()
        Log.d("OtaManager", "Downloaded version: $versionContent")

        return versionContent
    }

    /**
     * Clears the stored OTA data (path, version, download URL, and version check URL) from shared preferences.
     */
    fun clearStoredData() {
        preferences.clearOtaData()
        Log.d("OtaManager", "Cleared stored OTA data from SharedPreferences")
    }
}
